src/equations/bbm_1d.jl

```diff
@@ -0,0 +1,196 @@
+#   A Broad Class of Conservative Numerical Methods for Dispersive Wave Equations
+#   [DOI: 10.4208/cicp.OA-2020-0119](https://doi.org/10.4208/cicp.OA-2020-0119)
+function rhs!(dq, q, t, mesh, equations::BBMEquation1D, initial_condition,
+              ::BoundaryConditionPeriodic, source_terms, solver, cache)
+    (; invImD2, eta2, eta2_x, eta_x, etaeta_x) = cache
+    if solver.D1 isa PeriodicUpwindOperators
+        (; eta_x_upwind) = cache
+    end
+
+    eta, = q.x
+    deta, = dq.x
+    # energy and mass conservative semidiscretization
+    @trixi_timeit timer() "hyperbolic" begin
+        @.. eta2 = eta^2
+        if solver.D1 isa PeriodicDerivativeOperator ||
+           solver.D1 isa UniformPeriodicCoupledOperator ||
+           solver.D1 isa FourierDerivativeOperator
+            mul!(eta2_x, solver.D1, eta2)
+            mul!(eta_x, solver.D1, eta)
+            @.. etaeta_x = eta * eta_x
+            # split form
+            @.. deta = -(1 / 3 * (eta2_x + etaeta_x) + eta_x)
+        elseif solver.D1 isa PeriodicUpwindOperators
+            mul!(eta2_x, solver.D1.central, eta2)
+            mul!(eta_x_upwind, solver.D1.minus, eta)
+            mul!(eta_x, solver.D1.central, eta)
+            @.. etaeta_x = eta * eta_x
+            # split form
+            @.. deta = -(1 / 3 * (eta2_x + etaeta_x) + eta_x_upwind)
+        else
+            @error "unknown type of first-derivative operator: $(typeof(solver.D1))"
+        end
+    end
+
+    @trixi_timeit timer() "source terms" calc_sources!(dq, q, t, source_terms, equations,
+                                                       solver)
+
+    @trixi_timeit timer() "elliptic" begin
+        solve_system_matrix!(deta, invImD2, equations)
+    end
+    return nothing
+end
+
+"""
+    energy_total_modified!(e, q, equations::BBMEquation1D, cache)
+
+Return the modified total energy `e` of the primitive variables `q_global` for the
+[`BBMEquation1D`](@ref). The `energy_total_modified`
+is a conserved quantity (for periodic boundary conditions).
+
+It is given by
+```math
+\\frac{1}{2} \\eta^2 + \\frac{1}{2} \\eta_{xx}.
+```
+
+`q_global` is a vector of the primitive variables at ALL nodes.
+`cache` needs to hold the SBP operators used by the `solver`.
+
+See also [`energy_total_modified`](@ref).
+"""
+function energy_total_modified!(e, q_global, equations::BBMEquation1D, cache)
+    eta, = q_global.x
+
+    (; D1, D2, eta_xx, tmp1) = cache
+    if D1 isa PeriodicUpwindOperators
+        mul!(tmp1, D1.minus, eta)
+        mul!(eta_xx, D1.plus, tmp1)
+    else
+        mul!(eta_xx, D2, eta)
+    end
+
+    @.. e = 0.5 * (eta^2 + eta_xx)
+    return nothing
+end
+
+"""
+    invariant_cubic(q, equations::BBMEquation1D)
+
+Return the cubic invariant of the primitive variables `q` for the
+[`BBMEquation1D`](@ref). The cubic invariant is given by
+```math
+(\\eta + 1)^3.
+```
+"""
+function invariant_cubic(q, equations::BBMEquation1D)
```

Comment:
I'm not sure about the name `invariant_cubic`. Do you have another idea, @ranocha?

Reply:
It's a good question how to call this. Instead of the plain cubic invariant, we could also use the Hamiltonian, see https://doi.org/10.1007/s10543-023-00992-w
Then, it could make sense to implement the two semidiscretizations conserving either the quadratic invariant or the Hamiltonian containing cubic terms

Reply:
Makes sense. I'll do that. When we want to implement both semis, what would be a good way for an interface to switch between the two versions. We could add an argument to the equation to switch between the two semis. This would be easy to implement, but this is question of the discretization and not of the (physical) equation. So I think it doesn't really make sense to put this into the equation. We could also put it into the `solver`, but then how should we store the additional variable? Should it always be in a `solver` argument, also for other equations, where we don't need such a variable? This feels not very elegant. Any ideas?

Reply:
Well, the docstrings of the equations we have so far mention the conserved quantities of the semidiscretizations explicitly, so it does indeed make sense to put this choice into the equations - although this looks a bit strange from a physical point of view.

Reply:
Ok, so let's put it in the equation. It's maybe not ideal, but I can live with it. And if documented well enough it should also not be too confusing for a user.

Reply:
I'm thinking of how to implement the Hamiltonian. I would like to implement it in the usual form of a function, which only gets a local `q` and the `equations`. This is possible since it doesn't depend on any spatial derivatives in the case of the BBM equation. But is it save to assume that any Hamiltonian function we will implement in the future will not depend on any spatial derivatives? Because if it would, then we would need to use the interface we have for `energy_total_modified`, which gets the global `q_global` and the `cache` holding the SBP operator(s). I am, e.g., thinking of the cubic invariant of the Camassa-Holm equation $\int(u^3 + uu_x^2)$. Would we call this Hamiltonian? If yes, I would opt for already implementing the global variant for the Hamiltonian now and treat it similarly as the modified energy/entropy, such that we don't need to change that in the future. What are your thoughts on that @ranocha?

Reply:
There can be Hamiltonians that depend on derivatives of the solution. So we will need the more general interface

Reply:
I implemented the Hamiltonian in [577303c](https://github.com/JoshuaLampert/DispersiveShallowWater.jl/pull/150/commits/577303c0e80da745d673078dbd29ea905fab1243) and also added a new parameter `split_form` to switch between the two semidiscretizations.

Reply:
👍